### PR TITLE
Wip/scala client

### PR DESCRIPTION
**Pull request by johnnyaug**

### CODE REVIEW

File: clients/src/main/scala/io/treeverse/clients/s3/S3MetaRangeReader.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package io.treeverse.clients.s3
+
+import com.amazonaws.services.s3.AmazonS3
+import com.amazonaws.services.s3.model.GetObjectRequest
+import io.treeverse.clients
+import io.treeverse.clients.{MetaRangeReader, SSTableReader}
+
+import java.io.File
+
+class S3MetaRangeReader(var s3: AmazonS3, var bucket: String, var repositoryPrefix: String) extends MetaRangeReader {
```
Comment (ozkatz) [inline clients/src/main/scala/io/treeverse/clients/s3/S3MetaRangeReader.scala:10]: I don't know what's the final purpose of this client, but it *probably* shouldn't be calling S3 directly (i.e. if it's a Hadoop embeddable library, you should be coordinating IO through the HadoopFileSystem

File: clients/src/main/scala/io/treeverse/clients/VarInt.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * from https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/util/Varint.java
+ */
+object VarInt {
+  @throws[IOException]
+  def readSignedVarLong(in: DataInput): Long = {
+    val raw = readUnsignedVarLong(in)
+    // This undoes the trick in writeSignedVarLong()
+    val temp = (((raw << 63) >> 63) ^ raw) >> 1
```
Comment (ozkatz) [inline clients/src/main/scala/io/treeverse/clients/VarInt.scala:15]: love it.

File: clients/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+target
```
Comment (ozkatz) [inline clients/.gitignore:1]: I feel strongly that the Scala client should live in another repository and not in the main lakeFS one.

**Review by ozkatz — CHANGES_REQUESTED** [review]

File: clients/src/main/scala/io/treeverse/clients/VarInt.scala (hunk 2 above)
Comment (arielshaqed) [reply clients/src/main/scala/io/treeverse/clients/VarInt.scala:15]: A classic.  Only needed in JVMs and similar "unsigned is too difficult".  Zigzag usually looks something like this (e.g. in Go):
```go
	ux := uint64(x) << 1
	if x < 0 {
		ux = ^ux
	}
```
which has the advantage of not having to pick between `>>` and `>>>` while remaining branch-free...

File: clients/.gitignore (hunk 3 above)
Comment (arielshaqed) [reply clients/.gitignore:1]: I somewhat disagree: I have found it difficult to see the advantages of multi-repo.  But in any case a prerequisite for going multi-repo on our clients is to be able to publish our protobuf files to _somewhere_.  Saying "git submodules" does not tend to make friends...

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

A mixture of nits and important comments.  These are the important ones:
1. In JVM, strings are not the same as byte sequences.
2. Possible late dynamic loading; possibly unnecessary restriction of SSTable format to Snappy compression.
3. Exceptions should be typed.
4. Avoid mixing operations on 2 overloaded streams (if possible).

File: clients/project/protoc.sbt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+addSbtPlugin("com.thesamet" % "sbt-protoc" % "1.0.0")
+
+libraryDependencies += "com.thesamet.scalapb" %% "compilerplugin" % "0.10.10"
```
[the remark this answers is not in the document]
Comment (johnnyaug) [reply clients/project/protoc.sbt:3]: No, here is a dependency for the compile time, while the other one if for runtime

**Comment by johnnyaug** [comment]
Moved to a separate repository

**Pull request closed without merging**